storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java

```diff
@@ -13,286 +13,332 @@
-      // comply with Zipkin API as service names are required to be ordered lexicographically and store returns unordered values
-      Collections.sort(spanNames);
-      return spanNames;
+    @Override protected List<String> doExecute() throws IOException {
+      StreamsMetadata metadata =
+          traceStoreStream.metadataForKey(SPAN_NAMES_STORE_NAME, serviceName,
+              new StringSerializer());
+      HttpClient httpClient = httpClient(metadata);
+      AggregatedHttpResponse response =
+          httpClient.get(String.format("/service-names/%s/span-names", serviceName))
+              .aggregate()
+              .join();
+      if (!HttpStatus.OK.equals(response.status())) return new ArrayList<>();
+      String content = response.contentUtf8();
+      Set<String> set = GSON.fromJson(content, Set.class);
+      return new ArrayList<>(set);
+    }
+
+    @Override protected void doEnqueue(Callback<List<String>> callback) {
+      try {
+        callback.onSuccess(doExecute());
+      } catch (IOException e) {
+        callback.onError(e);
+      }
     }
 
     @Override public Call<List<String>> clone() {
-      return new GetSpanNamesCall(spanNamesStore, serviceName);
+      return new GetSpanNamesCall(traceStoreStream, serviceName);
     }
   }
 
-  static class GetRemoteServiceNamesCall extends KafkaStreamsStoreCall<List<String>> {
-    final ReadOnlyKeyValueStore<String, Set<String>> remoteServiceNamesStore;
+  static class GetRemoteServiceNamesCall extends Call.Base<List<String>> {
+    static final Gson GSON = new Gson();
+
+    final KafkaStreams traceStoreStream;
     final String serviceName;
 
-    GetRemoteServiceNamesCall(ReadOnlyKeyValueStore<String, Set<String>> remoteServiceNamesStore,
-        String serviceName) {
-      this.remoteServiceNamesStore = remoteServiceNamesStore;
+    GetRemoteServiceNamesCall(KafkaStreams traceStoreStream, String serviceName) {
+      this.traceStoreStream = traceStoreStream;
       this.serviceName = serviceName;
     }
 
-    @Override public List<String> query() {
-      if (serviceName == null) return new ArrayList<>();
-      Set<String> remoteServiceNamesSet = remoteServiceNamesStore.get(serviceName);
-      if (remoteServiceNamesSet == null) return new ArrayList<>();
-      List<String> remoteServiceNames = new ArrayList<>(remoteServiceNamesSet);
-      // comply with Zipkin API as service names are required to be ordered lexicographically
-      Collections.sort(remoteServiceNames);
-      return remoteServiceNames;
+    @Override protected List<String> doExecute() throws IOException {
+      StreamsMetadata metadata =
+          traceStoreStream.metadataForKey(REMOTE_SERVICE_NAMES_STORE_NAME, serviceName,
+              new StringSerializer());
+      HttpClient httpClient = httpClient(metadata);
+      AggregatedHttpResponse response =
+          httpClient.get(String.format("/service-names/%s/remote-service-names", serviceName))
```

Comment:
armeria "might" have a template for clients.. not sure.

Reply:
For templating, armeria relies on its integration with retrofit (`armeria-retrofit`). Probably not worth it, but it would remove the need for `String.format` and manual json parsing so can think about it.

Reply:
would template work even when hosts are dynamic? kafka streams instances might get moved around and initial list of instances could change.